libdnf/dnf-context.cpp

```diff
@@ -3531,10 +3531,8 @@ dnf_context_module_install(DnfContext * context, const char ** module_specs, GEr
                             auto matching = latest->getProfiles(profileName);
                             profiles.insert(profiles.begin(), matching.begin(), matching.end());
                         }
-                        // Fallback to the new `default` knob in the modulemd proper.
-                        // This isn't actually used yet, at least in Fedora.
                         if (profiles.empty()) {
-                            profiles.push_back(latest->getDefaultProfile());
+                            throw std::runtime_error("No default profile found for " + latest->getFullIdentifier());
```

Comment:
I see that `throw std::runtime_error` is using in the method several times. But this is C API and with GError ** error, therefore I would prefer that it will directly sets GError. At the end it sets GError because it catches  `std::runtime_error` and transforms it into GError. What's your opinion?

Reply:
The reason we `throw` and `catch` here is to match dnf's semantics of processing all the inputs before throwing an error. This also matches what functions like `dnf_context_module_enable` do, where messages are queued up during each iteration if we encounter errors. (And actually, this is what we're doing here too, where in the `catch` block we queue up a message.)

Reply:
I am really sorry, I get it now. You throw `std::runtime_error` because you don't want to set g_error (`std::runtime_error` in not catched by following macro). 
```
#define CATCH_TO_GERROR(RET)                                                   \
catch (const libdnf::Goal::Error& e) {                                       \
    g_set_error_literal(error, DNF_ERROR, e.getErrCode(), e.what());           \
    return RET;                                                                \
} catch (const libdnf::Error& e) {                                             \
    g_set_error_literal(error, DNF_ERROR, DNF_ERROR_INTERNAL_ERROR, e.what()); \
    return RET;                                                                \
}
```
But this is a wrong approach. `gboolean dnf_context_module_install()` is C API function therefore it must not thow any C++ exception. For people that uses C++ it is not a problem because the can catch it, but for C users it is unsolvable.


I propose to replace `throw std::runtime_error("No default profile found for " + latest->getFullIdentifier());`
by
```
g_set_error_literal();
return FALSE; 
```

Reply:
It is true that `dnf_context_module_enable` trow an error, but it happen only when the internal logic of the code would be incorrect. Therefore I expect to terminate program without possibility to catch it when it happen. It is possible to use assert() instead.

Reply:
> But this is a wrong approach. `gboolean dnf_context_module_install()` is C API function therefore it must not thow any C++ exception. For people that uses C++ it is not a problem because the can catch it, but for C users it is unsolvable.

The exceptions we throw here never actually bubble to the caller. They're immediately caught and converted to messages.